include/ringmesh/basic/frame.h

```diff
@@ -152,6 +152,80 @@ namespace RINGMesh
     };
     ALIAS_2D_AND_3D( ReferenceFrame );
 
+    template < index_t DIMENSION >
+    class RINGMESH_API FrameManipulator
+    {
+    public:
+        static vecn< DIMENSION > coords_from_global_to_frame(
+            const ReferenceFrame< DIMENSION >& reference_frame,
+            const vecn< DIMENSION >& global_coords )
+        {
+            GEO::Matrix< DIMENSION, double > base_change_matrix;
+            for( auto i : range( DIMENSION ) )
+            {
+                for( auto j : range( DIMENSION ) )
+                {
+                    base_change_matrix( i, j ) = reference_frame[j][i];
+                }
+            }
+            base_change_matrix = base_change_matrix.inverse();
+            vecn< DIMENSION > local_coords;
+            for( auto i : range( DIMENSION ) )
+            {
+                for( auto j : range( DIMENSION ) )
+                {
+                    local_coords[i] +=
+                        base_change_matrix( i, j )
+                        * ( global_coords[j] - reference_frame.origin()[j] );
+                }
+            }
+            return local_coords;
+        }
+
+        static vecn< DIMENSION > coords_from_frame_to_global(
+            const ReferenceFrame< DIMENSION >& reference_frame,
+            const vecn< DIMENSION >& local_coords )
+        {
+            vecn< DIMENSION > global_coords = reference_frame->origin();
+            for( auto coord : range( DIMENSION ) )
+            {
+                for( auto coor : range( DIMENSION ) )
+                {
+                    global_coords[coord] +=
+                        local_coords[coor] * reference_frame[coor][coord];
+                }
+            }
+            return global_coords;
+        }
+
+        static ReferenceFrame< DIMENSION > reference_frame_from_global_to_local(
+            const ReferenceFrame< DIMENSION >& reference_frame )
+        {
+            ReferenceFrame< DIMENSION > inverse_reference_frame;
```

Comment:
To avoid code duplication you can call above functions

Reply:
Functions above give the coordinates from specified frame to global or from global to specified frame, i don't really see how i could call them here without it being slower and unefficient ... (same as using this functions for functions above)

Reply:
Ok

Reply:
block 163-171 and block 205-213 are identical. Please remove code dupplication.